### PR TITLE
Display Git branch in Side Bar

### PR DESCRIPTION
**Pull request by petternorman**
First of all, thanks for this extension, it's a simple way to be more productive when handling multiple projects.
I missed the ability to see which git branch a folder is currently using so I fixed it locally. I saw this issue #378 and thought that I might as well push this PR in case anyone else also need this feature. To keep it simple and fast I opted to not try to fix showing the current git changes since that would require slower Git commands.

## Summary

Adds the ability to display the current Git branch next to project names in the Side Bar. This feature helps users quickly identify which branch each project is on without having to open the project.

Screenshot 2026-01-11 at 13 25 38

### Added
- New utility function `getGitBranch()` in `src/utils/git.ts` that:
  - Reads the current Git branch from `.git/HEAD` file
  - Handles Git worktrees (where `.git` is a file pointing to the actual Git directory)
  - Supports both absolute and relative paths for worktree references
  - Returns `undefined` for non-Git repositories, detached HEAD state, or errors
- New configuration setting `projectManager.showGitBranchInSideBar` (default: `false`)
- Git branch is displayed as detail text in the Side Bar when the setting is enabled
- Automatic refresh of branch information when:
  - The setting is changed
  - The window gains focus (to update branch info after switching branches)
- Test suite covering:
  - Regular Git repositories
  - Different branch names
  - Detached HEAD state
  - Git worktrees (absolute and relative paths)
  - Non-Git directories
  - Edge cases and error handling

### Updated
- `StorageProvider` to read and display Git branch information
- Extension configuration change handler to refresh when setting changes
- Window state change handler to refresh branch info on focus
- README.md with documentation for the new setting
- Localization files with translations for the new setting description

## Technical Details

The implementation reads the Git branch directly from the `.git/HEAD` file, avoiding the need for external Git commands. This approach:
- Works offline
- Is fast and lightweight
- Handles edge cases like worktrees gracefully
- Fails silently for non-Git projects

## Testing

All tests pass, including:
- Unit tests for `getGitBranch()` function
- Integration with StorageProvider
- Configuration change handling
- Window focus refresh behavior

## Documentation

- README.md updated with the new setting
- New setting localized with (AI-)translations

### CODE REVIEW

**Review by alefragnani — CHANGES_REQUESTED** [review]
Hi @petternorman ,

First of all, thank you for your kind words and PR. I’m glad the extension is useful, and hope you had a good experience building and debugging it.

About the PR, based on the original issue, I was expecting two additional features:
1. The _branch_ be visible also in the `Git` view
2. The _branch_ be visible also in the `Command Palette`

Specifically about the `Git` view, I wonder how this information would appear, if `projectManager.showParentFolderInfoOnDuplicates` is enabled, since it is displayed in the same `detail` property. Should be one or another? Should it combine `branch` + `baseFolder`? 

About the `Command Palette` we have a similar question. Today it displays the `path` in the `detail` attribute. Should then replace to display the `branch` instead? Or should also concatenate? I tend to option **b**, on both scenarios, at least for the initial release.

Other than that, the changes where as expected. Following localization usage, updated README and loading branch info using git files instead of git commands, for the exact reasons you mentioned. 🤝 

Thank you

**Comment by alefragnani** [comment]
About the PR itself, I reviewed the changes and it seems ok, but it is now presenting conflict with the master branch. So, I would ask you to resolve the conflicts (rebase) so I could properly test it.

Thank you

**Comment by petternorman** [comment]
The branch is rebased and the conflict is fixed. So please test it when you have time @alefragnani

**Review by alefragnani — CHANGES_REQUESTED** [review]
Hi @petternorman ,

I could properly test it and it is working just fine. I just noticed that scenarios in the `onDidChangeWindowState` event handler. Other than that, it is really good. 

Thank you!

**Review by alefragnani — CHANGES_REQUESTED** [review]
I just found a blocking issue in the Command Palette. Please take a look.

Thank you

**Review by chefmarvinimana-cyber — COMMENTED** [review]
Need ro have break

**Comment by tmsrise** [comment]
Just tried this extension out, and was disappointed I couldn't see the projects' git branch in the list. Would be really grateful if this was fixed up and merged!

**Comment by thenickg** [comment]
@petternorman @alefragnani — I hit the "invalid path" issue locally and put together a small patch that addresses both open review threads. Opened it against this PR's branch: petternorman/vscode-project-manager#1

Summary:

**`projectsPicker.ts`** — branch was concatenated into `description`, which the picker's `onDidChangeSelection` later reads as `rootPath`, producing the "invalid path" prompt on open. Moved branch into a separate `detail` field with the `$(git-branch)` codicon; `description` stays as the raw path.

**`providers.ts`** — `onDidChangeWindowState` now gates on `state.focused`, sidebar visibility, and the `git.showBranchName` setting; also refreshes `gitProvider` in addition to `storageProvider`, per the review comment above.

Diff (2 files, +8/-5):

```diff
$(git diff upstream/master..fix/branch-overview-review-feedback -- src/)
```

Happy to adjust if either of you prefers a different shape — e.g. a different icon or keeping `matchOnDetail` consistent.

**Comment by sim186** [comment]
I would really love to see this PR merged! Happy to test it if necessary!

**Comment by petternorman** [comment]
Apologies for taking so long to respond—life, work, and summer got in the way.

A huge thank-you to @thenickg for stepping in and addressing the outstanding review comments. I’ve now brought those changes into this PR branch.

@alefragnani, I hope this brings the PR close to a ready state. When you have time, I’d really appreciate another look and any guidance on changes that may still be needed.

Thank you both for your patience and help!

File: src/quickpick/projectsPicker.ts [hunk 1]
```diff
@@ -135,10 +136,18 @@ export async function pickProjects(projectStorage: ProjectStorage, locators: Loc
                         commands.executeCommand("setContext", "inProjectManagerList", true);
 
                         //
+                        const showGitBranchInCommandPalette = workspace.getConfiguration("projectManager").get<string>("git.showBranchName", "never");
                         folders = (<any[]>folders).map(folder => {
+                            let description = folder.description;
+                            if (showGitBranchInCommandPalette === "onlyInCommandPalette" || showGitBranchInCommandPalette === "always") {
+                                const gitBranch = getGitBranch(PathUtils.expandHomePath(folder.description));
+                                if (gitBranch) {
+                                    description = `${folder.description} | ${gitBranch}`;
```
[the remark this answers is not in the document]
Comment (alefragnani) [reply src/quickpick/projectsPicker.ts]: The fix was applied in https://github.com/alefragnani/vscode-project-manager/pull/884/commits/6ab478116447b7ca5f901218e20e9da3cbda3b0a by @thenickg and @petternorman

**Review by alefragnani — APPROVED** [review]

File: src/sidebar/providers.ts [hunk 2]
```diff
@@ -77,7 +77,13 @@ export class Providers {
...
             this.storageTreeView.onDidCollapseElement(async event => {
                 await this.handleStorageTreeViewExpansionChange(event, "collapsed");
-            })
+            }),
+            // Refresh git branch info when window gains focus and sidebar is visible
+            vscode.window.onDidChangeWindowState(async () => {
+                if (this.storageTreeView.visible) {
+                    this.storageProvider.refresh();
+                }
+            }),
```
[the remark this answers is not in the document]
Comment (alefragnani) [reply src/sidebar/providers.ts]: The fix was applied in https://github.com/alefragnani/vscode-project-manager/commit/6ab478116447b7ca5f901218e20e9da3cbda3b0a by @thenickg and @petternorman

**Comment by alefragnani** [comment]
Hi @petternorman and @thenickg ,

Thanks for addressing the latest comments. It seems to be working just fine now, and should be available in the next release, which should be out in the next couple of weeks.

Stay tuned.
